### PR TITLE
Ruby on Rails Project: Bare Metal Forms and Helpers

### PR DESCRIPTION
**Pull request by alvarorf**
In this pull request, we add the necessary code for the third project in the Ruby on Rails curriculum. The main purpose of this project was to get to know how to work with helper functions in Rails, specifically in the context of forms. Users can be created and edited with this form (a more thorough description may be found in the Readme).

Thank you for any errors that you may find and for any suggestions that you may have.

### CODE REVIEW

File: app/views/users/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<%= form_for :user, url: '/users', method: :post do |form| %>
```
Comment (maelfosso) [inline app/views/users/edit.html.erb:1]: uses `users_path` helper for action

File: app/views/users/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    <input type="password" name="user[password]">
+
+    <input type="hidden" name="authenticity_token" value="<%= form_authenticity_token %>">
+
+    <!-- other inputs here -->
+    <input type="submit" value="Submit This Form">
+</form>
+<% end %>
+
+<%= form_for :user, url: '/users', method: :post do |form| %>
```
Comment (maelfosso) [inline app/views/users/new.html.erb:14]: uses `users_path` helper for action

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+class UsersController < ApplicationController
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      redirect_to '/users/new'
```
Comment (maelfosso) [inline app/controllers/users_controller.rb:5]: Kindly, uses `new_user_path` helper

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required 🔴 

Hello,

Thank you for the well-done job so far.
However, there are still some issues 

1- Kindly fix the linters' errors, please

2- There is one missing form implementation, the one with `form_tag`, as required [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_form_tag2.png)
Kindly add it and comment it out.

3- As requested [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/edit_user.png), it should be possible to edit a user. 
Kindly add it, please.

4- Read the comments left below

Kindly, make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

File: config/routes.rb [hunk 4]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
+  root to: 'users#new'
+  resource :users, only: [:new, :create, :edit]
```
Comment (blackpintz) [inline config/routes.rb:3]: **Your `config/routes.rb` is missing an `:update` action.**

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  def index
+    @users = User.all
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      flash[:notice] = "User '#{@user.username}' has been created!"
+      sleep(2.5)
+      redirect_to edit_users_url
```
Comment (blackpintz) [inline app/controllers/users_controller.rb]: In the create action, you should redirect to `new_user_path` if a user is successfully saved.

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      sleep(2.5)
+      redirect_to new_users_url
+    else
+      render :edit
+    end
+  end
+
+  def new; end
+
+  def edit; end
```
Comment (blackpintz) [inline app/controllers/users_controller.rb:29]: In the edit action, you need to define a method that finds a user from the db using the `id` params.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    else
+      render :new
+    end
+  end
+
+  def update
+    if @user.update(user_params)
+      flash[:notice] = "User '#{@user.username}' has been updated!"
+      sleep(2.5)
+      redirect_to new_users_url
```
Comment (blackpintz) [inline app/controllers/users_controller.rb]: 1). You haven't defined the `@user` variable in the update action. Remember to use the find method and `id` params to determine the `@user` from your db.

2). redirect_to `edit_user_path` if the user is successfully saved.

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  <%= label_tag 'username' %>
+  <%= text_field_tag(:username) %><br/>
+  <%= label_tag 'email' %>
+  <%= text_field_tag(:email) %><br/>
+  <%= label_tag 'password' %>
+  <%= password_field_tag 'password' %><br/>
+  <%= submit_tag 'Create a new user'%>
+<% end %>
+<% end %>
+
```
Comment (blackpintz) [inline app/views/users/new.html.erb:20]: 1). The `if` statement on line one is not necessary.

2). The above-highlighted section(from line 2 to 20) should be commented out.

3). Apply the above two changes in the `edit.html.erb ` file.

**Review by blackpintz — CHANGES_REQUESTED** [review]
# Status: Changes Required

### Hello team,

**Great job implementing the changes. However, there are still a few fixes needed.**

1). **I am not able to edit a user. The proper edit route is `localhost:3000/users/:id/edit` and not `localhost:3000/users/new`. Every user is assigned an id when they are created in the database. That is the id you use in the edit route.** 

Example
```
user saved in the DB.
 => #<ActiveRecord::Relation [#<User id: 1, username: "Rina", email: "rina@gmail.com", password: [FILTERED], created_at: "2020-11-04 18:29:51", updated_at: "2020-11-04 18:29:51">]> 
The edit route for the above user should be:
 localhost:3000/users/1/edit
```
2). **Make corrections in your readme file after you configure the right edit route. It currently says that one should visit `localhost:3000/users/new` if they want to edit a user.**

3). **Check more comments below.**

### Happy coding 😃.

**PS: You can reach me on slack(Rose Wanjohi).**

**Review by rOluochKe — APPROVED** [review]
Hello Team
### Well done, you got it right :clap: The project has been approved. Congratulations :boom: :100:
### Merge and submit the project. Good luck with the next one.
- Happy coding!

**Pull request merged**